### PR TITLE
SONARJAVA-1972 - S2095 Allow users to specify types to be excluded fr…

### PR DESCRIPTION
**Pull request by fergarrui**
* Added a comma separated list as a parameter to exclude types in the check.

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java [hunk 1]
```diff
@@ -72,9 +81,21 @@
...
   };
+  
+  private static final List<String> excludedTypesList = new ArrayList<>();
+  
   private static final MethodMatcher[] CLOSEABLE_EXCEPTIONS = new MethodMatcher[] {
     MethodMatcher.create().typeDefinition("java.nio.file.FileSystems").name("getDefault").withoutParameter()
   };
+  
+  @Override
+  public void init(MethodTree methodTree, CFG cfg) {
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java]: Please do not override the `init()` method. Symbolic execution based checks are sensitive, and in this case we don't want to touch anything in the initialization phase.

The best approach in order to retrieve the list of excluded resource types is to use a dedicated (private) getter, which lazily initialize the list (which then should not be static, with all the side effects you noticed).

File: java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java [hunk 2]
```diff
@@ -116,34 +137,39 @@ private static boolean needsClosing(Type type) {
         return false;
       }
     }
+    for (String excludedType : excludedTypesList) {
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java]: If the list is `static`, then you don't have to remove the static modifier everywhere. However, I would prefer you to use a dedicated getter lazily loading the configuration of excluded types.

File: java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java [hunk 3]
```diff
@@ -36,27 +44,28 @@
...
 @Rule(key = "S2095")
 public class UnclosedResourcesCheck extends SECheck {
 
   private enum Status {
     OPENED, CLOSED
   }
 
+  private static final String DEFAULT_EXCLUDED_TYPES = "";
+  
+  @RuleProperty(
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java]: I updated the description of the rule in order to include the parameter. Please follow the description from here for keys and description: [RSPEC-3591](https://jira.sonarsource.com/browse/RSPEC-3591)

File: java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java [hunk 4]
```diff
@@ -72,9 +81,21 @@
     "com.sun.org.apache.xml.internal.security.utils.UnsyncByteArrayOutputStream",
     "org.springframework.context.ConfigurableApplicationContext"
   };
+  
+  private static final List<String> excludedTypesList = new ArrayList<>();
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java]: Put this list (which should not be static) next to the other rule property-related field. Grouping will help.

File: java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java [hunk 5]
```diff
@@ -116,34 +137,39 @@ private static boolean needsClosing(Type type) {
...
+    for (String excludedType : excludedTypesList) {
+      if (type.is(excludedType)) {
+        return false;
+      }
+    }
     return isCloseable(type);
   }
 
-  private static boolean isCloseable(Type type) {
+  private boolean isCloseable(Type type) {
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java]: I'm not sure you need to remove all of these `static` modifiers. Some of them are probably still valid (this one notably)

File: java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java [hunk 6]
```diff
@@ -19,7 +19,15 @@
  */
 package org.sonar.java.se.checks;
 
+import java.util.ArrayList;
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java]: Please discard these changes, and only limit the modifications to the new imports.

**Comment by fergarrui** [comment]
Hi, I did the changes request:
* Discarded the changes in the imports (I'm using Eclipse formatter suggested by Sonar, that's why they were reordered, I'll keep in mind not to reorder them again).
* Rule properties updated to match [RSPEC-3591](https://jira.sonarsource.com/browse/RSPEC-3591)
* Grouped parameter-related properties. Also removed static modifier from the list holding the excluded types.
* Removed init() override and added a lazy load to populate the list.
* Added back some static modifiers, but I can't add most of them.

Thanks for your suggestions and patience! Please let me know if you want more improvements/changes.

File: java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java [hunk 7]
```diff
@@ -116,29 +125,43 @@ private static boolean needsClosing(Type type) {
...
+    if ( excludedTypesList.isEmpty() && !StringUtils.isBlank(excludedTypes)) {
+      for (String excludedType : excludedTypes.split(",")) {
+        excludedTypesList.add(excludedType.trim());
+      }
+    }
+    return excludedTypesList;
+  }
 
-  private static boolean isCloseable(Type type) {
+  private boolean isCloseable(Type type) {
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java]: This one can still be static.

File: java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java [hunk 8]
```diff
@@ -116,29 +125,43 @@ private static boolean needsClosing(Type type) {
...
-  private static boolean isOpeningResource(NewClassTree syntaxNode) {
+  private boolean isOpeningResource(NewClassTree syntaxNode) {
     if (isWithinTryHeader(syntaxNode)) {
       return false;
     }
     return needsClosing(syntaxNode.symbolType());
   }
 
-  private static boolean isWithinTryHeader(Tree syntaxNode) {
+  private boolean isWithinTryHeader(Tree syntaxNode) {
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java]: And when `isTryStatementResource` is static, this one can be static as well.

File: java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java [hunk 9]
```diff
@@ -116,29 +125,43 @@ private static boolean needsClosing(Type type) {
...
     final Tree parent = syntaxNode.parent();
     if (parent.is(Tree.Kind.VARIABLE)) {
       return isTryStatementResource((VariableTree) parent);
     }
     return false;
   }
 
-  private static boolean isTryStatementResource(VariableTree variable) {
+  private boolean isTryStatementResource(VariableTree variable) {
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java]: This one can be static.

**Comment by fergarrui** [comment]
Changes done :)

**Pull request merged**